Classes/Extensions/DDFileLogger+Buffering.m

```diff
@@ -0,0 +1,155 @@
+//   to endorse or promote products derived from this software without specific
+//   prior written permission of Deusty, LLC.
+
+#import "DDFileLogger+Buffering.h"
+#import "DDFileLogger+Internal.h"
+
+static NSUInteger kMaximumBytesCountInBuffer = (1 << 10) * (1 << 10); // 1 MB.
+static NSUInteger kDefaultBytesCountInBuffer = (1 << 10);
+
+// MARK: Public Interface
+@interface DDBufferedProxy<FileLogger: DDFileLogger *>: NSProxy
++ (instancetype)decoratedInstance:(FileLogger)instance;
+@property (assign, nonatomic, readwrite) NSUInteger maximumBytesCountInBuffer;
+@end
+
+@interface DDBufferedProxy<FileLogger: DDFileLogger *> () {
+    NSOutputStream *_bufferStream;
+    NSUInteger _bufferSize;
+}
+- (instancetype)initWithInstance:(FileLogger)instance;
+@property (strong, nonatomic, readwrite) FileLogger instance;
+@end
+
+@interface DDBufferedProxy (StreamManipulation)
+- (void)flushBuffer;
+- (void)dumpBufferToDisk;
+- (void)appendToBuffer:(NSData *)data;
+- (BOOL)isBufferFull;
+@end
+
+@implementation DDBufferedProxy (StreamManipulation)
+
+- (void)flushBuffer {
+    // do something.
+    [_bufferStream close];
+    _bufferStream = nil;
+    _bufferSize = 0;
+}
+
+- (void)dumpBufferToDisk {
+    // do something.
+    __auto_type data = (NSData *)[_bufferStream propertyForKey:NSStreamDataWrittenToMemoryStreamKey];
+    [self.instance logData:data];
+    [self flushBuffer];
+}
+
+- (void)appendToBuffer:(NSData *)data {
+    __auto_type length = data.length;
+    if (data.length != 0) {
+        if (_bufferStream == nil) {
+            _bufferStream = [[NSOutputStream alloc] initToMemory];
+            [_bufferStream open];
+            _bufferSize = 0;
+        }
+        __auto_type appendedData = (const uint8_t *)calloc(length, sizeof(uint8_t));
+        [data getBytes:(void *)appendedData length:length];
+        if (appendedData != NULL) {
+            [_bufferStream write:appendedData maxLength:length];
+        }
+        if (appendedData != NULL) {
+            free((void *)appendedData);
+        }
+        _bufferSize += length;
+    }
+}
+
+- (BOOL)isBufferFull {
+    return _bufferSize > self.maximumBytesCountInBuffer;
+}
+
+@end
+
+@implementation DDBufferedProxy
+@synthesize maximumBytesCountInBuffer = _maximumBytesCountInBuffer;
+
+#pragma mark - Properties
+- (void)setMaximumBytesCountInBuffer:(NSUInteger)maximumBytesCountInBuffer {
+    _maximumBytesCountInBuffer = MIN(maximumBytesCountInBuffer, kMaximumBytesCountInBuffer);
+}
+
+#pragma mark - Initialization
++ (instancetype)decoratedInstance:(DDFileLogger *)instance {
+    return [[self alloc] initWithInstance:instance];
+}
+
+- (instancetype)initWithInstance:(DDFileLogger *)instance {
+    self.instance = instance;
+    self.maximumBytesCountInBuffer = kDefaultBytesCountInBuffer;
+    return self;
+}
+
+- (void)dealloc {
+    [self dumpBufferToDisk];
+    self.instance = nil;
+}
+
+#pragma mark - Logging
+- (void)logData:(NSData *)data {
+    if ([self isBufferFull]) {
```

Comment:
This check would be better done after the `appendToBuffer` so that it gets flushed immediately when full rather than wait for the next `logData` call.